### PR TITLE
Redo Dilithium/clean and add Dilithium's AVX2 implementations

### PR DESCRIPTION
**Pull request by thomwiggers**
I've cleaned up Dilithium's source code again and am now also adding the AVX2 code.

#### Manually checked properties

* [x] No stringification macros
* [x] Output-parameter pointers in functions are on the left
* [x] Negative return values on failure of API functions (within restrictions of FO transform).
* [x] `const` arguments are labeled as `const`
* [x] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [x] All integer types are of fixed size, using `stdint.h` types (including `uint8_t` instead of `unsigned char`)
  * [x] Integers used for indexing are of size `size_t`

### CODE REVIEW

**Comment by thomwiggers** [comment]
I'm currently still looking at if I can get this running on MacOS via https://github.com/pq-crystals/dilithium/pull/10

File: crypto_sign/dilithium2/clean/ntt.c [hunk 1]
```diff
@@ -113,25 +114,25 @@ void PQCLEAN_DILITHIUM2_CLEAN_ntt(uint32_t p[N]) {
 *
 * Arguments:   - uint32_t p[N]: input/output coefficient array
 **************************************************/
-void PQCLEAN_DILITHIUM2_CLEAN_invntt_frominvmont(uint32_t p[N]) {
-    unsigned int start, len, j, k;
+void PQCLEAN_DILITHIUM2_CLEAN_invntt_frominvmont(uint32_t *p) {
```
Comment (mkannwischer) [inline crypto_sign/dilithium2/clean/ntt.c]: What's the reason for changing these?

File: crypto_sign/dilithium2/clean/packing.c [hunk 2]
```diff
@@ -4,188 +4,186 @@
...
+* Name:        PQCLEAN_DILITHIUM2_CLEAN_pack_sig
 *
 * Description: Bit-pack signature sig = (z, h, c).
 *
-* Arguments:   - unsigned char sig[]: output byte array
+* Arguments:   - uint8_t sig[]: output byte array
 *              - const polyvecl *z: pointer to vector z
 *              - const polyveck *h: pointer to hint vector h
-*              - const poly *c: pointer to challenge polynomial
+*              - const poly *c: pointer to PQCLEAN_DILITHIUM2_CLEAN_challenge polynomial
```
Comment (mkannwischer) [inline crypto_sign/dilithium2/clean/packing.c]: ?

File: crypto_sign/dilithium2/clean/packing.c [hunk 3]
```diff
@@ -208,48 +206,49 @@ void PQCLEAN_DILITHIUM2_CLEAN_pack_sig(unsigned char sig[CRYPTO_BYTES],
...
-* Name:        unpack_sig
+* Name:        PQCLEAN_DILITHIUM2_CLEAN_unpack_sig
 *
 * Description: Unpack signature sig = (z, h, c).
 *
 * Arguments:   - polyvecl *z: pointer to output vector z
 *              - polyveck *h: pointer to output hint vector h
-*              - poly *c: pointer to output challenge polynomial
-*              - const unsigned char sig[]: byte array containing
+*              - poly *c: pointer to output PQCLEAN_DILITHIUM2_CLEAN_challenge polynomial
```
Comment (mkannwischer) [inline crypto_sign/dilithium2/clean/packing.c]: ?

File: crypto_sign/dilithium2/clean/sign.h [hunk 4]
```diff
@@ -1,30 +1,12 @@
...
+#define PQCLEAN_DILITHIUM2_CLEAN_SIGN_H
 
+#include "api.h"
 #include "params.h"
 #include "poly.h"
 #include "polyvec.h"
 
-void PQCLEAN_DILITHIUM2_CLEAN_expand_mat(polyvecl mat[K], const unsigned char rho[SEEDBYTES]);
-void PQCLEAN_DILITHIUM2_CLEAN_challenge(poly *c, const unsigned char mu[CRHBYTES],
+void PQCLEAN_DILITHIUM2_CLEAN_expand_mat(polyvecl mat[K], const uint8_t rho[SEEDBYTES]);
```
Comment (mkannwischer) [inline crypto_sign/dilithium2/clean/sign.h]: and why no change it here?

File: crypto_sign/dilithium2/clean/ntt.c (hunk 1 above)
Comment (thomwiggers) [reply crypto_sign/dilithium2/clean/ntt.c]: This was still `uint32_t *p` in the upstream code.

File: crypto_sign/dilithium2/clean/packing.c (hunk 2 above)
Comment (thomwiggers) [reply crypto_sign/dilithium2/clean/packing.c]: Probably got renamed by a refactoring operation (there's a function called `challenge` in `sign.c`).

File: crypto_sign/dilithium2/clean/ntt.c (hunk 1 above)
Comment (thomwiggers) [reply crypto_sign/dilithium2/clean/ntt.c]: Ah, no, weird, this was present in the upstream code. Not sure how this got changed.

**Pull request merged**